### DISCUSSION TITLE
embedding request failed with status code '401' - You didn't provide an API key despite API_KEY being set

### DISCUSSION THREAD

**Discussion by rodrigomeireles** [meysamhadeli/codai#73] [Q&A] [answered]
I have both OPENAI_API_KEY and API_KEY set. I tried passing it as a CLI argument as well. I tried adding it to the `config.yml`. I tried combinations of those.

Is there any step I am missing?

**Comment by meysamhadeli** [comment]
> I have both OPENAI_API_KEY and API_KEY set. I tried passing it as a CLI argument as well. I tried adding it to the `config.yml`. I tried combinations of those.
> 
> Is there any step I am missing?

@rodrigomeireles Your api key for embedding and chat is the same or different!? you try with openai? if use openai chat should use embeding by openai and if use ollama chat should use embedding by ollama. 

- If your key in chat and embedding is different it not support yet and please create a issue for that and I will add it soon. otherwise if the key are the same it should work.

**Reply by rodrigomeireles** [reply]
Sorry, can you give me a full list of the necessary keys I need to set up? The README.md mentions only API_KEY but you're talking about embedding keys as well.

I only set up API_KEY.  I am using Open AI.

**Reply by meysamhadeli** [reply]
In simple scenario 1 key is enough if the server of chat and embedding is the same. but for example If we use azure openai and server of chat and embedding are in different server location, the keys are different and here if keys for each one is different for chat and embedding, the app doesn't support yet. I will add support for different keys today.

- If you are using OpenAI and the Key for chat and embedding is the same app should work properly.

**Comment by rodrigomeireles** [comment]
Then I believe this is an issue. I'm running on OSC M3, my environment
variable is set and I can't run it.

**Reply by meysamhadeli** [reply]
> Then I believe this is an issue. I'm running on OSC M3, my environment variable is set and I can't run it.
> […](#)
> On Sat, 16 Nov 2024, 09:48 Meysam Hadeli, ***@***.***> wrote: In simple scenario 1 key is enough if the server of chat and embedding is the same. but for example If we use azure openai and server of chat and embedding are in different server location, the keys are different and here if keys for each one is different for chat and embedding, the app doesn't support yet. I will add support for different keys today. - If you are using OpenAI and the Key for chat and embedding is the same app should work properly. — Reply to this email directly, view it on GitHub <[#73 (reply in thread)](https://github.com/meysamhadeli/codai/discussions/73#discussioncomment-11277342)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJQUPSM5VW46HM4ZLTHMNDD2A45LRAVCNFSM6AAAAABR4X7BLKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRXG4ZTIMQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

@rodrigomeireles Could you try to pass your key in terminal with this `codai code --api_key "your_api_key"` and check what's happen!? and delete `API_KEY` from environment. (I guess it has a problem with reading key from environment variable!?)

Also I created a Issue #74 to support different keys for chat and embedding.

**Reply by meysamhadeli** [reply]
@rodrigomeireles Now, I implemented supporting different keys for chat and embeddings. . #76

**Reply by rodrigomeireles** [reply]
Hi @meysamhadeli sorry for the dealy, I was traveling. I did exactly as you instructed, deleted de API_KEY env var and passed the CLI argument for the key and got the same error. This time I even tried using Bash (I use Nushell as default) and I got the same response.

**Reply by meysamhadeli** [reply, marked as answer]
@rodrigomeireles Hi, Don't worry. I figure out openai provider had problem with configure token in the code and I fixed it last day. could you try with latest version of codai `1.7.2` and see the problem fixed or not!? (also, I simplified the config and please follow with new config base on doc.)

> please let me know it fixed or not with latest version!?

**Reply by rodrigomeireles** [reply]
This was fixed in the latest version.

**Reply by rodrigomeireles** [reply]
By the way, this was an issue.

**Reply by meysamhadeli** [reply]
Thanks for informing me to fix the issue :)
